### PR TITLE
jit.script OutputLayer and trace source/target embedding during inference

### PR DESCRIPTION
**Pull request by fhieber**
#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/layers_pt.py [hunk 1]
```diff
@@ -101,17 +102,16 @@ def __init__(self,
             self.weight = weight
         self.bias = pt.nn.Parameter(pt.Tensor(vocab_size))
 
-        self.previous_slice_ids = None  # type: Optional[pt.Tensor]
-        self.reduced_weight_bias = None  # type: Optional[Tuple[pt.Tensor, pt.Tensor]]
+        self.previous_slice_ids: Optional[pt.Tensor] = pt.empty(0)
+        self.reduced_weight: Optional[pt.Tensor] = pt.empty(0)
+        self.reduced_bias: Optional[pt.Tensor] = pt.empty(0)
```
Comment (mjdenkowski) [inline sockeye/layers_pt.py:107]: Do these still need to be `Optional` (or have type annotations at all) now that they're created as `Tensor`s?

**Review by mjdenkowski — COMMENTED** [review]
It looks like this still needs a version/CHANGELOG update.

Other than that and the minor typing comment, everything should be ready to merge.

File: sockeye/layers_pt.py (hunk 1 above)
Comment (fhieber) [reply sockeye/layers_pt.py:107]: Good catch, we can remove the type annotation.

File: sockeye/encoder_pt.py [hunk 2]
```diff
@@ -90,14 +90,17 @@ class PyTorchEmbedding(PyTorchEncoder):
...
-
         if embedding is not None:
             self.embedding = embedding
         else:
             self.embedding = pt.nn.Embedding(self.config.vocab_size, self.config.num_embed,
                                              sparse=self.config.allow_sparse_grad)
 
+        self.num_factors = self.config.num_factors
+        if self.num_factors > 1:
+            assert len(self.config.factor_configs) == self.num_factors - 1
```
Comment (tdomhan) [inline sockeye/encoder_pt.py:101]: Do we expect this to be violated?
If we want to keep it should it be converted to `check_condition` + a message?

**Review by mjdenkowski — APPROVED** [review]

File: sockeye/encoder_pt.py (hunk 2 above)
Comment (fhieber) [reply sockeye/encoder_pt.py:101]: No, I think its safe to assume this is always true. I will remove the assertion.

**Pull request merged**